### ISSUE TITLE
Grabbing Keys Requires Special privilidges

### ISSUE THREAD

**Issue by TTWNO** [odilia-app/odilia#8]
It is not ideal to require special permissions (and dangerous ones, at that) to be able to snoop keyboard events for a screen reader.

Other, more responsible screen readers, may use the official `atspi` method of capturing keys. There are a couple problems with this:

1. Keys will only be forwarded to the screen reader when an application is already accessible.
2. The long process of being passed from the kernel, to `libinput` to the window manager/DE, to an application, over DBus, finally to a screen reader is not very efficient.

## Keys Only For "Good" Applications

Unless I am completely mistaken, and I very well might be (I encourage criticism): the action of forward keys to the screen reader is required to be implemented by the application or toolkit used for application development. Therefore, if there is some bug in the application which causes key events to not read, or if the application is not accessible at all, the user now has no ability to control their screen reader, resulting in a state where that application must be killed from a console for the user to continue using their computer normally.
I might be wrong, but I think somebody had this problem with Zoom on Linux recently. Maybe it's better now, or that wasn't the problem, but it sounded suspiciously familiar.

## Long Process

In addition to essentially being optional for an application to send keys, even in the case that keys are forwarded properly, the latency can be a problem on low-power devices like a Raspberry Pi, or Pinebook Pro; areas which Linux normally excels in, but the accessibility infrastructure isn't suited for. To be clear, this is not any one person or organization's fault, just the way things are at this time.

## Solution?

At this time, Odilia is interrupting keys at the kernel level, deciding whether to consume them, or pass them through to the rest of the input stack.

This has some *very* major disadvantages of its own.

1. It required access to a privileged interface (`evdev`, `uinput`), which causes *serious* security concerns among the more paranoid (myself included).
2. It does *not* respect keymaps, or language settings. When using a European or Russian keyboard, Odilia will still only be bale to bind to the English letter it *thinks* is represented by the incoming keycode. This is not ideal whatsoever.

At this time, it seems like the only solution, and I would certainly rather a very fast, annoying to configure screen reader than one which is slow/buggy but easy to configure.
Perhaps this bridge can be mended one day, but it would require two things to happen:

1. Allow a11y programs to capture key input *before* it is sent to the application (X11). (Side note: how would sudo prompts work? I see why this is complicated.)
2. Allow a11y programs to capture key input *before* it is sent to the application (Wayland). Same side not above stands.

Wayland could likely be extended with an a11y protocol extension. I'm not sure what the right idea with X11 may be, or if there is any existing solution to this problem on X.

Put your ideas here. Let's find a way that works best for everyone. A performant, secure screen reader is our goal, and we won't stop until we've got it.

**Comment by TTWNO** [comment]
Partially addressed by https://github.com/flatpak/xdg-desktop-portal/pull/711

**Comment by TTWNO** [comment]
Going to close this on merge of #99
